src/librustc/diagnostics.rs

```diff
@@ -353,26 +353,35 @@ If you really want global mutable state, try using `static mut` or a global
 "##,
 
 E0018: r##"
-The value of static and const variables must be known at compile time. You
-can't cast a pointer as an integer because we can't know what value the
-address will take.
-
-However, pointers to other constants' addresses are allowed in constants,
-example:
+The value of static and constant integers must be known at compile time. You
+can't cast a pointer to an integer because the address of a pointer can
+vary.
 
+For example, if you write:
 ```
-const X: u32 = 50;
-const Y: *const u32 = &X;
+static MY_STATIC: u32 = 42;
+static MY_STATIC_ADDR: usize = &MY_STATIC as *const _ as usize;
+static WHAT: usize = (MY_STATIC_ADDR^17) + MY_STATIC_ADDR;
 ```
 
-Therefore, casting one of these non-constant pointers to an integer results
-in a non-constant integer which lead to this error. Example:
+Then MY_STATIC_ADDR would contain the address of MY_STATIC. However,
+that address can change when the program is linked, as well as change
```

Comment:
"that address", since we're refering to a precise address, isn't "this" more recommended?